### PR TITLE
Fix wait logic

### LINKED ISSUE #174: Change Speak trait interface
Current

```rust
pub trait Speaker: Send + Sync {
    fn speak(&self, message: &str);
}
```

Proposed

```rust
pub trait Speaker: Send + Sync {
    async fn speak(&self, message: &str) -> Result<()>;
}
```

It should be `async` because it takes sometime to finish speaking.
And it might be failed.

### LINKED ISSUE #228: Go back to sync
It seems that async is not used effectively,  and it has weak reason to use async for now.
Let's go back to old implementation.

Return boxed Wait instead of Future.

```rust
use crate::error::Error;

pub type Wait<'a> = Box<dyn WaitTrait + 'a>;

pub trait WaitTrait {
    fn wait(&self) -> Result<(), Error>;
}
```

Current

```rust
a.send_joint_positions()?.await?;
```

Go back to

```rust
a.send_joint_positions()?.wait()?;
```

### LINKED ISSUE #245: Do not wait in joint_positions_sender and teleop
It's better not to wait in some apps.

Blocked by #228

### PR DESCRIPTION
**Pull request by taiki-e**
- Make CompleteCondition::wait async fn 
- Change async methods to normal methods that return future 
- Do not wait in teleop and gui
- Change Speak trait interface to return result and future

Closes #228
Closes #245
Closes #174

### CODE REVIEW

File: arci/src/waits.rs [hunk 1]
```diff
@@ -1,12 +1,80 @@
...
+    pub fn from_stream(stream: impl Stream<Item = Result<(), Error>> + Send + 'a) -> Self {
+        Self::new(async move {
+            futures::pin_mut!(stream);
+            while stream.try_next().await?.is_some() {}
+            Ok(())
+        })
+    }
+
+    /// Creates a new `WaitFuture` which immediately complete.
+    pub fn dummy() -> Self {
```
[the remark this answers is not in the document]
Comment (taiki-e) [reply arci/src/waits.rs:55]: Sounds good. I'll change it to `ready`.

File: openrr-teleop/src/joints.rs [hunk 2]
```diff
@@ -127,6 +127,7 @@ where
                 };
             self.joint_trajectory_client
                 .send_joint_positions(pos, self.step_duration)
+                .unwrap()
                 .await
                 .unwrap();
```
Comment (OTL) [inline openrr-teleop/src/joints.rs:132]: Below change solve the teleop problem.
I check it macos + urdf-viz, but I believe that it works on Linux/ROS.

```suggestion
            // do not wait
            let _ = self.joint_trajectory_client
                .send_joint_positions(pos, self.step_duration)
                .unwrap();
```

File: openrr-teleop/src/joints.rs [hunk 3]
```diff
@@ -79,8 +79,11 @@ where
             GamepadEvent::ButtonPressed(Button::East) => {
                 self.joint_index = (self.joint_index + 1) % self.dof;
                 self.submode = format!("{}", self.joint_index);
-                self.speaker
-                    .speak(&format!("{}{}", self.mode, self.submode()));
+                // do not wait
+                let _ = self
+                    .speaker
+                    .speak(&format!("{}{}", self.mode, self.submode()))
+                    .unwrap();
```
Comment (taiki-e) [inline openrr-teleop/src/joints.rs:86]: Should we wait for this speak?

File: openrr-teleop/src/joints_pose_sender.rs [hunk 4]
```diff
@@ -79,8 +79,11 @@ where
                 self.pose_index = (self.pose_index + 1) % self.joints_poses.len();
                 let joints_pose = &self.joints_poses[self.pose_index];
                 self.submode = format!(" {} {}", joints_pose.client_name, joints_pose.pose_name);
-                self.speaker
-                    .speak(&format!("{}{}", self.mode, self.submode()));
+                // do not wait
+                let _ = self
+                    .speaker
+                    .speak(&format!("{}{}", self.mode, self.submode()))
+                    .unwrap();
```
Comment (taiki-e) [inline openrr-teleop/src/joints_pose_sender.rs:86]: Same here.

File: openrr-teleop/src/joints.rs (hunk 3 above)
Comment (OTL) [reply openrr-teleop/src/joints.rs:86]: no wait here. (Keep your change)

File: openrr-teleop/src/joints_pose_sender.rs (hunk 4 above)
Comment (OTL) [reply openrr-teleop/src/joints_pose_sender.rs:86]: no wait here. (Keep your change)

**Review by OTL — COMMENTED** [review]
In details, we can fix after merge. let's go!

**Review by OTL — APPROVED** [review]

**Pull request merged**